debitpolicy/debitpolicy.go

```diff
@@ -0,0 +1,115 @@
+package debitpolicy
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/dogmatiq/dogma"
+	"github.com/dogmatiq/example/messages/commands"
+	"github.com/dogmatiq/example/messages/events"
+)
+
+// debitpolicy is the aggregate root for the debit policy.
+type debitpolicy struct {
+	// DebitAmount is the current sum of debit amounts for the period, in cents.
+	DebitAmount int64
+}
+
+func (p *debitpolicy) ApplyEvent(m dogma.Message) {
+	switch x := m.(type) {
+	case events.AccountDebitedForWithdrawal:
+		p.DebitAmount += x.Amount
+	case events.AccountDebitedForTransfer:
+		p.DebitAmount += x.Amount
+	}
+}
+
+// Aggregate implements the business logic for a debit policy.
+//
+// It centralizes all debits that are applied to an account over a period of
+// time in order to enforce a limited debit policy.
+type Aggregate struct{}
+
+// New returns a new debit policy instance.
+func (Aggregate) New() dogma.AggregateRoot {
+	return &debitpolicy{}
+}
+
+// Configure configures the behaviour of the engine as it relates to this
+// handlers.
+func (Aggregate) Configure(c dogma.AggregateConfigurer) {
+	c.Name("debitpolicy")
+	c.RouteCommandType(commands.CheckWithdrawalAllowedByDebitPolicy{})
+	c.RouteCommandType(commands.CheckTransferAllowedByDebitPolicy{})
+}
+
+// RouteCommandToInstance returns the ID of the aggregate instance that is
+// targetted by m.
+func (Aggregate) RouteCommandToInstance(m dogma.Message) string {
+	switch x := m.(type) {
+	case commands.CheckWithdrawalAllowedByDebitPolicy:
+		return makeInstanceID(x.TransactionTimestamp, x.AccountID)
+	case commands.CheckTransferAllowedByDebitPolicy:
+		return makeInstanceID(x.TransactionTimestamp, x.AccountID)
+	default:
+		panic(dogma.UnexpectedMessage)
+	}
+}
+
+// HandleCommand handles a command message that has bene routed to this handler.
+func (Aggregate) HandleCommand(s dogma.AggregateCommandScope, m dogma.Message) {
+	switch x := m.(type) {
+	case commands.CheckWithdrawalAllowedByDebitPolicy:
+		checkWithdrawalAllowed(s, x)
+	case commands.CheckTransferAllowedByDebitPolicy:
+		checkTransferAllowed(s, x)
+	default:
+		panic(dogma.UnexpectedMessage)
+	}
+}
+
+func checkWithdrawalAllowed(s dogma.AggregateCommandScope, m commands.CheckWithdrawalAllowedByDebitPolicy) {
+	s.Create()
+
+	p := s.Root().(*debitpolicy)
+
+	if p.DebitAmount+m.Amount <= debitLimitPerPeriod {
```

Comment:
I just wanted to point out as well, that it's ok to add methods in the aggregate root.  So for example, instead of doing this math here and again for transfers, you could call something like `p.IsAmountWithinLimit(m.Amount)`, further clarifying the intent.

Reply:
irrc, that's very analogous to an unexposed method we have in hive, deploy w/f, computing if all settings reference have been downloaded or not.